# Lab | Data Structuring and Combining Data

## Challenge 1: Combining & Cleaning Data

In this challenge, we will be working with the customer data from an insurance company, as we did in the two previous labs. The data can be found here:
- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv

But this time, we got new data, which can be found in the following 2 CSV files located at the links below.

- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file2.csv
- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file3.csv

Note that you'll need to clean and format the new data.

Observation:
- One option is to first combine the three datasets and then apply the cleaning function to the new combined dataset
- Another option would be to read the clean file you saved in the previous lab, and just clean the two new files and concatenate the three clean datasets

In [61]:
import pandas as pd

In [2]:
# Your code goes here
df1 = pd.read_csv("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv")

In [3]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4008 entries, 0 to 4007
Data columns (total 11 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Customer                   1071 non-null   object 
 1   ST                         1071 non-null   object 
 2   GENDER                     954 non-null    object 
 3   Education                  1071 non-null   object 
 4   Customer Lifetime Value    1068 non-null   object 
 5   Income                     1071 non-null   float64
 6   Monthly Premium Auto       1071 non-null   float64
 7   Number of Open Complaints  1071 non-null   object 
 8   Policy Type                1071 non-null   object 
 9   Vehicle Class              1071 non-null   object 
 10  Total Claim Amount         1071 non-null   float64
dtypes: float64(3), object(8)
memory usage: 344.6+ KB


In [4]:
df1.rename(columns={'ST': 'Customer_state'}, inplace=True)

In [5]:
df1.rename(columns={'GENDER': 'Gender'}, inplace=True)

In [6]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4008 entries, 0 to 4007
Data columns (total 11 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Customer                   1071 non-null   object 
 1   Customer_state             1071 non-null   object 
 2   Gender                     954 non-null    object 
 3   Education                  1071 non-null   object 
 4   Customer Lifetime Value    1068 non-null   object 
 5   Income                     1071 non-null   float64
 6   Monthly Premium Auto       1071 non-null   float64
 7   Number of Open Complaints  1071 non-null   object 
 8   Policy Type                1071 non-null   object 
 9   Vehicle Class              1071 non-null   object 
 10  Total Claim Amount         1071 non-null   float64
dtypes: float64(3), object(8)
memory usage: 344.6+ KB


In [7]:
df2 = pd.read_csv("/Users/admin/Desktop/IRONHACK/2.WEEK/LABS_2/lab-dw-data-structuring-and-combining/raw.githubusercontent.com_data-bootcamp-v4_data_main_file2.csv")

In [8]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 996 entries, 0 to 995
Data columns (total 11 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Customer                   996 non-null    object 
 1   ST                         996 non-null    object 
 2   GENDER                     991 non-null    object 
 3   Education                  996 non-null    object 
 4   Customer Lifetime Value    992 non-null    object 
 5   Income                     996 non-null    int64  
 6   Monthly Premium Auto       996 non-null    int64  
 7   Number of Open Complaints  996 non-null    object 
 8   Total Claim Amount         996 non-null    float64
 9   Policy Type                996 non-null    object 
 10  Vehicle Class              996 non-null    object 
dtypes: float64(1), int64(2), object(8)
memory usage: 85.7+ KB


In [9]:
df2.rename(columns={'ST': 'Customer_state'}, inplace=True)
# I will rename columns Gender to avoid duplicate the columns
df2.rename(columns={'GENDER': 'Gender'}, inplace=True)

In [10]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 996 entries, 0 to 995
Data columns (total 11 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Customer                   996 non-null    object 
 1   Customer_state             996 non-null    object 
 2   Gender                     991 non-null    object 
 3   Education                  996 non-null    object 
 4   Customer Lifetime Value    992 non-null    object 
 5   Income                     996 non-null    int64  
 6   Monthly Premium Auto       996 non-null    int64  
 7   Number of Open Complaints  996 non-null    object 
 8   Total Claim Amount         996 non-null    float64
 9   Policy Type                996 non-null    object 
 10  Vehicle Class              996 non-null    object 
dtypes: float64(1), int64(2), object(8)
memory usage: 85.7+ KB


In [11]:
df3 = pd.read_csv("/Users/admin/Desktop/IRONHACK/2.WEEK/LABS_2/lab-dw-data-structuring-and-combining/raw.githubusercontent.com_data-bootcamp-v4_data_main_file3.csv")

In [12]:
df3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7070 entries, 0 to 7069
Data columns (total 11 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Customer                   7070 non-null   object 
 1   State                      7070 non-null   object 
 2   Customer Lifetime Value    7070 non-null   float64
 3   Education                  7070 non-null   object 
 4   Gender                     7070 non-null   object 
 5   Income                     7070 non-null   int64  
 6   Monthly Premium Auto       7070 non-null   int64  
 7   Number of Open Complaints  7070 non-null   int64  
 8   Policy Type                7070 non-null   object 
 9   Total Claim Amount         7070 non-null   float64
 10  Vehicle Class              7070 non-null   object 
dtypes: float64(2), int64(3), object(6)
memory usage: 607.7+ KB


In [13]:
df3.rename(columns={'State': 'Customer_state'}, inplace=True)

In [14]:
df3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7070 entries, 0 to 7069
Data columns (total 11 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Customer                   7070 non-null   object 
 1   Customer_state             7070 non-null   object 
 2   Customer Lifetime Value    7070 non-null   float64
 3   Education                  7070 non-null   object 
 4   Gender                     7070 non-null   object 
 5   Income                     7070 non-null   int64  
 6   Monthly Premium Auto       7070 non-null   int64  
 7   Number of Open Complaints  7070 non-null   int64  
 8   Policy Type                7070 non-null   object 
 9   Total Claim Amount         7070 non-null   float64
 10  Vehicle Class              7070 non-null   object 
dtypes: float64(2), int64(3), object(6)
memory usage: 607.7+ KB


In [15]:
new_df = pd.concat([df1, df2, df3], axis=0, keys=["Customer", "Customer", "Customer"])

In [16]:
new_df.info()

<class 'pandas.core.frame.DataFrame'>
MultiIndex: 12074 entries, ('Customer', 0) to ('Customer', 7069)
Data columns (total 11 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Customer                   9137 non-null   object 
 1   Customer_state             9137 non-null   object 
 2   Gender                     9015 non-null   object 
 3   Education                  9137 non-null   object 
 4   Customer Lifetime Value    9130 non-null   object 
 5   Income                     9137 non-null   float64
 6   Monthly Premium Auto       9137 non-null   float64
 7   Number of Open Complaints  9137 non-null   object 
 8   Policy Type                9137 non-null   object 
 9   Vehicle Class              9137 non-null   object 
 10  Total Claim Amount         9137 non-null   float64
dtypes: float64(3), object(8)
memory usage: 1.4+ MB


In [17]:
#create a dict to applied rename(). Create a dict: original value as key and new value as value.
new_df.rename(columns={'Customer':'customer',
                       'Customer_state': 'customer_state',
                       'Gender':'gender',
                       'Education':'education',
                       'Customer Lifetime Value':'customer_lifetime_value',
                       'Income':'income',
                       'Monthly Premium Auto':'monthly_auto_premium',
                       'Number of Open Complaints':'open_complaints',
                       'Policy Type':'policy_type',
                       'Vehicle Class':'vehicle_class',
                       'Total Claim Amount':'total_claim_amount' 
                  },
               inplace=True)

In [18]:
new_df.info()

<class 'pandas.core.frame.DataFrame'>
MultiIndex: 12074 entries, ('Customer', 0) to ('Customer', 7069)
Data columns (total 11 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   customer                 9137 non-null   object 
 1   customer_state           9137 non-null   object 
 2   gender                   9015 non-null   object 
 3   education                9137 non-null   object 
 4   customer_lifetime_value  9130 non-null   object 
 5   income                   9137 non-null   float64
 6   monthly_auto_premium     9137 non-null   float64
 7   open_complaints          9137 non-null   object 
 8   policy_type              9137 non-null   object 
 9   vehicle_class            9137 non-null   object 
 10  total_claim_amount       9137 non-null   float64
dtypes: float64(3), object(8)
memory usage: 1.4+ MB


In [23]:
pip install regex

Note: you may need to restart the kernel to use updated packages.


In [19]:
import re

`Column: gender`
- Gender column contains various inconsistent values such as "F", "M", "Femal", "Male", "female", which need to be standardized, for example, to "M" and "F".

In [22]:
new_df["gender"].unique()

array([nan, 'F', 'M', 'Femal', 'Male', 'female'], dtype=object)

In [23]:
new_df.info()

<class 'pandas.core.frame.DataFrame'>
MultiIndex: 12074 entries, ('Customer', 0) to ('Customer', 7069)
Data columns (total 11 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   customer                 9137 non-null   object 
 1   customer_state           9137 non-null   object 
 2   gender                   9015 non-null   object 
 3   education                9137 non-null   object 
 4   customer_lifetime_value  9130 non-null   object 
 5   income                   9137 non-null   float64
 6   monthly_auto_premium     9137 non-null   float64
 7   open_complaints          9137 non-null   object 
 8   policy_type              9137 non-null   object 
 9   vehicle_class            9137 non-null   object 
 10  total_claim_amount       9137 non-null   float64
dtypes: float64(3), object(8)
memory usage: 1.4+ MB


In [24]:
gender_dic = {'Femal': 'F',
              'Male': 'M',
              'female': 'F'}
new_df['gender'] = new_df['gender'].replace(gender_dic, regex=True)
new_df["gender"].unique()

array([nan, 'F', 'M'], dtype=object)

`Column: customer_state`
- State abbreviations be can replaced with its full name, for example "AZ": "Arizona", "Cali": "California", "WA": "Washington"

In [25]:
new_df.info()

<class 'pandas.core.frame.DataFrame'>
MultiIndex: 12074 entries, ('Customer', 0) to ('Customer', 7069)
Data columns (total 11 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   customer                 9137 non-null   object 
 1   customer_state           9137 non-null   object 
 2   gender                   9015 non-null   object 
 3   education                9137 non-null   object 
 4   customer_lifetime_value  9130 non-null   object 
 5   income                   9137 non-null   float64
 6   monthly_auto_premium     9137 non-null   float64
 7   open_complaints          9137 non-null   object 
 8   policy_type              9137 non-null   object 
 9   vehicle_class            9137 non-null   object 
 10  total_claim_amount       9137 non-null   float64
dtypes: float64(3), object(8)
memory usage: 1.4+ MB


In [26]:
duplicated_columns = new_df.columns[new_df.columns.duplicated()]
duplicated_columns

Index([], dtype='object')

In [27]:
new_df.info()

<class 'pandas.core.frame.DataFrame'>
MultiIndex: 12074 entries, ('Customer', 0) to ('Customer', 7069)
Data columns (total 11 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   customer                 9137 non-null   object 
 1   customer_state           9137 non-null   object 
 2   gender                   9015 non-null   object 
 3   education                9137 non-null   object 
 4   customer_lifetime_value  9130 non-null   object 
 5   income                   9137 non-null   float64
 6   monthly_auto_premium     9137 non-null   float64
 7   open_complaints          9137 non-null   object 
 8   policy_type              9137 non-null   object 
 9   vehicle_class            9137 non-null   object 
 10  total_claim_amount       9137 non-null   float64
dtypes: float64(3), object(8)
memory usage: 1.4+ MB


In [28]:
new_df["customer_state"].unique()

array(['Washington', 'Arizona', 'Nevada', 'California', 'Oregon', 'Cali',
       'AZ', 'WA', nan], dtype=object)

In [29]:
dict_customer_state  = {r"\b[Ww]ash(ington)?\b": "Washington",
                       r"\b[Aa]riz(ona)?\b": "Arizona",
                       r"\b[Nn]eva(da)?\b": "Nevada",
                       r"\b[Cc]ali(fornia)?\b": "California",
                       r"\bAZ\b": "Arizona",
                       r"\bWA\b": "Washington"
                       }
new_df['customer_state'] = new_df['customer_state'].replace(dict_customer_state, regex=True)
new_df["customer_state"].unique()

array(['Washington', 'Arizona', 'Nevada', 'California', 'Oregon', nan],
      dtype=object)

`Column: education`
- In education, "Bachelors" could be replaced by "Bachelor"

In [32]:
new_df['education'] = new_df['education'].str.replace('Bachelors', 'Bachelor')
new_df["education"].unique()

array(['Master', 'Bachelor', 'High School or Below', 'College', 'Doctor',
       nan], dtype=object)

`Column: lifetime_value`
- In Customer Lifetime Value, delete the % character

In [33]:
new_df["customer_lifetime_value"] = new_df["customer_lifetime_value"].str.replace(r'\s*%\s*', '', regex=True)
new_df["customer_lifetime_value"].unique()

array([nan, '697953.59', '1288743.17', ..., '568964.41', '368672.38',
       '399258.39'], dtype=object)

`Column: vehicle_class`
- In vehicle class, "Sports Car", "Luxury SUV" and "Luxury Car" could be replaced by "Luxury"

In [34]:
dict_vehicle_class = {'Sports Car':'Luxury',
                     'Luxury SUV': 'Luxury',
                     'Luxury Car':'Luxury'}
new_df['vehicle_class'] = new_df['vehicle_class'].replace(dict_vehicle_class, regex=True)
new_df["vehicle_class"].unique()

array(['Four-Door Car', 'Two-Door Car', 'SUV', 'Luxury', nan],
      dtype=object)

`Column: customer_lifetime_value`
- Customer lifetime value should be numeric

In [35]:
new_df["customer_lifetime_value"] = pd.to_numeric(new_df["customer_lifetime_value"])
new_df["customer_lifetime_value"].unique()

array([       nan,  697953.59, 1288743.17, ...,  568964.41,  368672.38,
        399258.39])

`Column: open_complaints`
- Number of open complaints has an incorrect format. Look at the different values it takes with unique() and take the middle value. As an example, 1/5/00 should be 5. Number of open complaints is a string - remember you can use split() to deal with it and take the number you need. Finally, since it should be numeric, cast the column to be in its proper type.

In [40]:
new_df['open_complaints'] = new_df['open_complaints'].str.split('/', expand=True)[1] #expand=True se agrega para manejar los nulls
new_df['open_complaints'] = pd.to_numeric(new_df["open_complaints"], errors='coerce')

AttributeError: Can only use .str accessor with string values!

In [41]:
new_df['open_complaints'].unique

<bound method Series.unique of Customer  0       0.0
          1       0.0
          2       0.0
          3       0.0
          4       0.0
                 ... 
          7065    NaN
          7066    NaN
          7067    NaN
          7068    NaN
          7069    NaN
Name: open_complaints, Length: 12074, dtype: float64>

In [39]:
new_df.info()

<class 'pandas.core.frame.DataFrame'>
MultiIndex: 12074 entries, ('Customer', 0) to ('Customer', 7069)
Data columns (total 11 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   customer                 9137 non-null   object 
 1   customer_state           9137 non-null   object 
 2   gender                   9015 non-null   object 
 3   education                9137 non-null   object 
 4   customer_lifetime_value  2060 non-null   float64
 5   income                   9137 non-null   float64
 6   monthly_auto_premium     9137 non-null   float64
 7   open_complaints          2067 non-null   float64
 8   policy_type              9137 non-null   object 
 9   vehicle_class            9137 non-null   object 
 10  total_claim_amount       9137 non-null   float64
dtypes: float64(5), object(6)
memory usage: 1.4+ MB


### Formating data type

In [48]:
#fill the null values with a specific value (such as the column mean or median for numerical variables, 
#and mode for categorical variables)

`Categorical variable:` for categorical variables I decide to apply the mode to fill nulls values, choosing the value that appears most frequently in that column and using it to replace the missing values:
- customer_state
- gender
- education
- policy_type
- vehicle_class

In [42]:
mode_customer_state = new_df['customer_state'].mode()
new_df['customer_state'].fillna('California', inplace=True)

In [44]:
mode_gender = new_df['gender'].mode()
new_df['gender'].fillna('F', inplace=True)

In [45]:
mode_education = new_df['education'].mode()
new_df['education'].fillna('Bachelor', inplace=True)

In [46]:
mode_policy_type = new_df['policy_type'].mode()
new_df['policy_type'].fillna('Personal Auto', inplace=True)

In [47]:
mode_vehicle_class = new_df['vehicle_class'].mode()
new_df['vehicle_class'].fillna('Four-Door Car', inplace=True)

In [49]:
new_df.isnull()

customer  customer_state  gender  education  \
Customer 0        False           False   False      False   
         1        False           False   False      False   
         2        False           False   False      False   
         3        False           False   False      False   
         4        False           False   False      False   
...                 ...             ...     ...        ...   
         7065     False           False   False      False   
         7066     False           False   False      False   
         7067     False           False   False      False   
         7068     False           False   False      False   
         7069     False           False   False      False   

               customer_lifetime_value  income  monthly_auto_premium  \
Customer 0                        True   False                 False   
         1                       False   False                 False   
         2                       False   False                 False   
         3                       False   False                 False   
         4                       False   False                 False   
...                                ...     ...                   ...   
         7065                     True   False                 False   
         7066                     True   False                 False   
         7067                     True   False                 False   
         7068                     True   False                 False   
         7069                     True   False                 False   

               open_complaints  policy_type  vehicle_class  total_claim_amount  
Customer 0               False        False          False               False  
         1               False        False          False               False  
         2               False        False          False               False  
         3               False        False          False               False  
         4               False        False          False               False  
...                        ...          ...            ...                 ...  
         7065             True        False          False               False  
         7066             True        False          False               False  
         7067             True        False          False               False  
         7068             True        False          False               False  
         7069             True        False          False               False  

[12074 rows x 11 columns]

`Numerical variables:` for numerical variables I decide to apply the median because is less sensitive to extreme values:
- customer
- customer_lifetime_value
- income
- monthly_auto_premium
- open_complaints
- total_claim_amount

In [50]:
new_df['customer'].fillna('none', inplace=True)

In [51]:
media_customer_lifetime = new_df['customer_lifetime_value'].median()
new_df['customer_lifetime_value'].fillna(588174.235, inplace=True)

In [52]:
median_income = new_df['income'].median()
new_df['income'].fillna(36234.0, inplace=True)

In [53]:
median_auto = new_df['monthly_auto_premium'].median()
print(median_auto)
new_df['monthly_auto_premium'].fillna(83.0, inplace=True)

83.0


In [57]:
new_df['open_complaints'].fillna(0, inplace=True)

In [58]:
new_df.isnull()

customer  customer_state  gender  education  \
Customer 0        False           False   False      False   
         1        False           False   False      False   
         2        False           False   False      False   
         3        False           False   False      False   
         4        False           False   False      False   
...                 ...             ...     ...        ...   
         7065     False           False   False      False   
         7066     False           False   False      False   
         7067     False           False   False      False   
         7068     False           False   False      False   
         7069     False           False   False      False   

               customer_lifetime_value  income  monthly_auto_premium  \
Customer 0                       False   False                 False   
         1                       False   False                 False   
         2                       False   False                 False   
         3                       False   False                 False   
         4                       False   False                 False   
...                                ...     ...                   ...   
         7065                    False   False                 False   
         7066                    False   False                 False   
         7067                    False   False                 False   
         7068                    False   False                 False   
         7069                    False   False                 False   

               open_complaints  policy_type  vehicle_class  total_claim_amount  
Customer 0               False        False          False               False  
         1               False        False          False               False  
         2               False        False          False               False  
         3               False        False          False               False  
         4               False        False          False               False  
...                        ...          ...            ...                 ...  
         7065            False        False          False               False  
         7066            False        False          False               False  
         7067            False        False          False               False  
         7068            False        False          False               False  
         7069            False        False          False               False  

[12074 rows x 11 columns]

# Challenge 2: Structuring Data

In this challenge, we will continue to work with customer data from an insurance company, but we will use a dataset with more columns, called marketing_customer_analysis.csv, which can be found at the following link:

https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis_clean.csv

This dataset contains information such as customer demographics, policy details, vehicle information, and the customer's response to the last marketing campaign. Our goal is to explore and analyze this data by performing data cleaning, formatting, and structuring.

In [63]:
# Your code goes here
mkt = pd.read_csv("/Users/admin/Desktop/IRONHACK/2.WEEK/LABS_2/lab-dw-data-structuring-and-combining/marketing_customer_analysis_clean.csv")

In [64]:
mkt.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10910 entries, 0 to 10909
Data columns (total 27 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   unnamed:_0                     10910 non-null  int64  
 1   customer                       10910 non-null  object 
 2   state                          10910 non-null  object 
 3   customer_lifetime_value        10910 non-null  float64
 4   response                       10910 non-null  object 
 5   coverage                       10910 non-null  object 
 6   education                      10910 non-null  object 
 7   effective_to_date              10910 non-null  object 
 8   employmentstatus               10910 non-null  object 
 9   gender                         10910 non-null  object 
 10  income                         10910 non-null  int64  
 11  location_code                  10910 non-null  object 
 12  marital_status                 10910 non-null 

In [68]:
mkt.head()

unnamed:_0 customer       state  customer_lifetime_value response  \
0           0  DK49336     Arizona              4809.216960       No   
1           1  KX64629  California              2228.525238       No   
2           2  LZ68649  Washington             14947.917300       No   
3           3  XL78013      Oregon             22332.439460      Yes   
4           4  QA50777      Oregon              9025.067525       No   

   coverage education effective_to_date employmentstatus gender  ...  \
0     Basic   College        2011-02-18         Employed      M  ...   
1     Basic   College        2011-01-18       Unemployed      F  ...   
2     Basic  Bachelor        2011-02-10         Employed      M  ...   
3  Extended   College        2011-01-11         Employed      M  ...   
4   Premium  Bachelor        2011-01-17    Medical Leave      F  ...   

   number_of_policies     policy_type        policy  renew_offer_type  \
0                   9  Corporate Auto  Corporate L3            Offer3   
1                   1   Personal Auto   Personal L3            Offer4   
2                   2   Personal Auto   Personal L3            Offer3   
3                   2  Corporate Auto  Corporate L3            Offer2   
4                   7   Personal Auto   Personal L2            Offer1   

   sales_channel  total_claim_amount  vehicle_class  vehicle_size  \
0          Agent          292.800000  Four-Door Car       Medsize   
1    Call Center          744.924331  Four-Door Car       Medsize   
2    Call Center          480.000000            SUV       Medsize   
3         Branch          484.013411  Four-Door Car       Medsize   
4         Branch          707.925645  Four-Door Car       Medsize   

  vehicle_type month  
0            A     2  
1            A     1  
2            A     2  
3            A     1  
4            A     1  

[5 rows x 27 columns]

1. You work at the marketing department and you want to know which sales channel brought the most sales in terms of total revenue. Using pivot, create a summary table showing the total revenue for each sales channel (branch, call center, web, and mail).
Round the total revenue to 2 decimal points.  Analyze the resulting table to draw insights.

In [69]:
mkt['sales_channel'].unique()

array(['Agent', 'Call Center', 'Branch', 'Web'], dtype=object)

In [77]:
def mean_round(x):
    return round(x.mean(), 2)

In [80]:
mkt.pivot_table(
    values="total_claim_amount",
    index="sales_channel",
    aggfunc=mean_round
)

total_claim_amount
sales_channel                    
Agent                      439.27
Branch                     430.58
Call Center                432.79
Web                        434.56

2. Create a pivot table that shows the average customer lifetime value per gender and education level. Analyze the resulting table to draw insights.

In [86]:
mkt.pivot_table(
    values="customer_lifetime_value",
    index=["gender","education"],
    aggfunc=mean_round
)

customer_lifetime_value
gender education                                    
F      Bachelor                              7874.27
       College                               7748.82
       Doctor                                7328.51
       High School or Below                  8675.22
       Master                                8157.05
M      Bachelor                              7703.60
       College                               8052.46
       Doctor                                7415.33
       High School or Below                  8149.69
       Master                                8168.83

**Gender Differences:**

- For female customers (F), the average customer lifetime value is higher for those with a "High School or Below" (8675.22) and "Master's" (8157.05) education level.
- For male customers (M), the average customer lifetime value is higher for those with a "Master's" (8168.83) and "College" (8052.46) education level.
- Overall, the "High School or Below" education level has the highest average customer lifetime value regardless of gender.
- Customers with a "Doctor" education level have lower average customer lifetime values compared to other education levels, both for males and females.
- Customers with a "College" education level also tend to have lower average customer lifetime values, although they are higher than values for those with a "Doctor" education level.

**Consistency Across Genders:**

For each education level, the average customer lifetime values do not vary significantly between males and females. Both genders exhibit similar trends in how education level affects the average customer lifetime value.

## Bonus

You work at the customer service department and you want to know which months had the highest number of complaints by policy type category. Create a summary table showing the number of complaints by policy type and month.
Show it in a long format table.

*In data analysis, a long format table is a way of structuring data in which each observation or measurement is stored in a separate row of the table. The key characteristic of a long format table is that each column represents a single variable, and each row represents a single observation of that variable.*

*More information about long and wide format tables here: https://www.statology.org/long-vs-wide-data/*

In [ ]:
# Your code goes here